### PR TITLE
Finalizing LXD remote credentials

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following change finalizes the LXD remote credentials. This PR
adds two different concepts:

 1. When adding credentials, it is sometimes preferable to finalize
 the credentials if you're supplying just a file. This is optional
 and doesn't require other buy in from existing providers.
 2. Using a trust.password for talking to a LXD server is defined
 as insecure[1], so we tend to not want to use it other than getting
 the server cert onto the local machine.

Additional changes include, the removal of interactive mode, this has
now been aligned with certificate authorisation type now that we have
the following code. Introduction of a manual authorization type, where
we can supply all the certificates at once.

[1] https://github.com/lxc/lxd/blob/master/doc/security.md#production-setup

## QA steps

The following steps can be used to verify that bootstrapping to a 
remote lxd server is possible, without manually exchanging certs.

#### Files:

```yaml
clouds:
    lxd:
        type: lxd
        auth-types: [certificate]
        endpoint: $LXD_SERVER_URL
```

```yaml
credentials:
    lxd:
        lxd:
            auth-type: certificate
            trust-password: $LXD_TRUST_PASSWORD
            client-cert: |
                $CLIENT_CERT
            client-key: |
                $CLIENT_KEY
```

#### Steps:

```sh
juju add-cloud lxd /path/to/cloud-file.yaml
juju add-credential lxd -f /path/to/credential-file.yaml
juju bootstrap lxd
```

## Documentation changes

You no longer require the manual certificate moving between host
and server.

This removes the interactive auth type, instead we use the 
`certificate` auth type. This type correctly parses the right files
and doesn't need the interactive work around employed in previous 
commits.

There also exists a `manual` auth type, which you can supply just
the client cert and key, along with the server cert. @mitechie might
have some thoughts on this, so maybe dropped in the future?

## Bug reference

N/A

### CODE REVIEW

**Comment by SimonRichardson** [comment]
https://trello.com/c/eWkNu6WH/295-use-lxd-cluster-api-to-ship-a-cert-to-use-for-communication

File: provider/lxd/credentials.go [hunk 1]
```diff
@@ -25,14 +29,15 @@ import (
...
-	interactiveAuthType = "interactive"
+	credAttrServerCert    = "server-cert"
+	credAttrClientCert    = "client-cert"
+	credAttrClientKey     = "client-key"
+	credAttrTrustPassword = "trust-password"
+
+	// manualAuthType is a credential auth-type provided as an option, where by
+	// you've already setup a lxd remote and you have all the certs at hand to
+	// just pass directly to juju.
+	manualAuthType = "manual"
```
Comment (manadart) [inline provider/lxd/credentials.go]: I think this is better how it was:
Certificate (I have all these certs) vs Interactive (I will generate them and do the handshake with the server via trust password).

Would current credentials be broken by this change? Not that it is that important on the dev branch.

File: provider/lxd/credentials.go [hunk 2]
```diff
@@ -219,36 +264,122 @@ func (p environProviderCredentials) finalizeCertificateCredential(
...
+	// We're not local, so setup the remote server and automate the remote
+	// certificate credentials.
+	return p.finalizeRemoteCertificateCredential(
+		stderr,
+		args.CloudEndpoint,
+		args.Credential,
+	)
+}
+
+func (p environProviderCredentials) finalizeRemoteCertificateCredential(
```
Comment (manadart) [inline provider/lxd/credentials.go]: Some of this looks similar to logic in container/lxd/certificate.go and _CreateClientCertificate_ in container/lxd/server.go.

If any of that code is now unused, we can delete it.

File: cmd/juju/cloud/addcredential.go [hunk 3]
```diff
@@ -180,6 +198,13 @@ func (c *addCredentialCommand) Run(ctxt *cmd.Context) error {
 	return nil
 }
 
+func (c *addCredentialCommand) shouldFinalizeCredential(provider environs.EnvironProvider, cred jujucloud.Credential) bool {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/juju/cloud/addcredential.go]: I have the exact same thought!

File: provider/lxd/credentials.go (hunk 1 above)
Comment (SimonRichardson) [reply provider/lxd/credentials.go]: I think this is the key to the whole PR, I'm not sure to answer your question. I'm also undecided what to do about this.

File: provider/lxd/credentials.go (hunk 2 above)
Comment (SimonRichardson) [reply provider/lxd/credentials.go]: OK, I'll have a look.

Comment (SimonRichardson) [reply provider/lxd/credentials.go]: fixed in https://github.com/juju/juju/pull/8938/commits/4edf5be9c5ec7206586fba1c3f22a0848d74f50e

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**